src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java

```diff
@@ -672,7 +699,10 @@ private void flush(final PartitionBuffer buff)
       String content = buff.getData();
       conn.putWithCache(stageName, fileName, content);
 
-      flushedOffset.set(Math.max(buff.getLastOffset() + 1, flushedOffset.get()));
+      // This is safe and atomic
+      flushedOffset.updateAndGet((value) -> Math.max(buff.getLastOffset() + 1, value));
+      pipeStatus.flushedOffset.set(flushedOffset.get() - 1);
```

Comment:
To make it clear:
processedOffset: a
flushedOffset: a + 1
committedOffset: a + 1

pipeStatus.processedOffset: a
pipeStatus.flushedOffset: a
pipeStatus.committedOffset: a
pipeStatus.purgedOffset: a